google/resumable_media/requests/download.py

```diff
@@ -160,13 +171,48 @@ def consume(
         if self._stream is not None:
             request_kwargs["stream"] = True
 
+        # Assign object generation if generation is specified in the media url.
+        if self._object_generation is None:
+            self._object_generation = _helpers._get_generation_from_url(self.media_url)
+
         # Wrap the request business logic in a function to be retried.
         def retriable_request():
+            url = self.media_url
+
+            # To restart an interrupted download, read from the offset of last byte
+            # received using a range request, and set object generation query param.
+            if self._bytes_downloaded > 0:
+                _download.add_bytes_range(
+                    self._bytes_downloaded, self.end, self._headers
+                )
+                request_kwargs["headers"] = self._headers
+
+                # Set object generation query param to ensure the same object content is requested.
+                if (
+                    self._object_generation is not None
+                    and _helpers._get_generation_from_url(self.media_url) is None
+                ):
+                    query_param = [("generation", self._object_generation)]
+                    url = _helpers.add_query_parameters(self.media_url, query_param)
+
             result = transport.request(method, url, **request_kwargs)
 
+            # If a generation hasn't been specified, and this is the first response we get, let's record the
+            # generation. In future requests we'll specify the generation query param to avoid data races.
+            if self._object_generation is None:
+                self._object_generation = _helpers._parse_generation_header(
+                    result, self._get_headers
+                )
+
             self._process_response(result)
 
+            # With decompressive transcoding, GCS serves back the whole file regardless of the range request,
+            # thus we reset the stream position to the start of the stream.
+            # See: https://cloud.google.com/storage/docs/transcoding#range
             if self._stream is not None:
+                if _helpers._is_decompressive_transcoding(result, self._get_headers):
+                    self._stream.seek(0)
```

Comment:
I didn't realize that the stream used for this method is seekable. Interesting!

Reply:
I ran a few experiments and the stream is seekable, though I'm not 100% sure if we can guarantee it to be seekable. 

The `requests.Response` object seems to be immutable, and [`iter_content()`](https://github.com/googleapis/google-resumable-media-python/blob/b529b780253bed9e5619e4c5eb2d612e975d874c/google/resumable_media/requests/download.py#L102) iterates from the beginning of the response. I wasn't able to find anything similar in Python to the io.CopyN approach in Go.

@frankyn @andrewsg @tseaver  Would love to hear your ideas on how to avoid writing duplicate data to the stream for decompressive transcoding interruptions. An example scenario could be:

--> initial download request
--> some bytes have already been written to stream
!!! connection error mid-stream
--> resumes download and sends a retried range request
--> the response serves the entire requested object, instead of the partial contents requested 
(decompressive transcoding does not support range requests)

Reply:
Sorry, to clarify, the remote server will _not_ reject a request with an unsupported range parameter, but instead will simply serve the wrong (complete) range?

What the heck? That sounds catastrophically broken

Reply:
I only see two solutions: just error out on seek() failing (with a try/except to transform the normal error message into something useful), or simply throw away all bytes up to tell() without writing them to the stream at all, and then write remaining bytes to the stream.

I prefer the simpler solution of the former, absent any customer demand for the more complex solution. A user deliberately passing in an un-seekable stream is a rare case and we don't need to write a bunch of hard to test code to support it unless it's obvious people are really doing that.

Reply:
> Sorry, to clarify, the remote server will _not_ reject a request with an unsupported range parameter, but instead will simply serve the wrong (complete) range?
> 
> What the heck? That sounds catastrophically broken

That's right... the server does _not_ reject the request. To quote the [doc](https://cloud.google.com/storage/docs/transcoding#range)

_When transcoding occurs, if the request for the object includes a Range header, that header is silently ignored._

Reply:
> I only see two solutions: just error out on seek() failing (with a try/except to transform the normal error message into something useful), or simply throw away all bytes up to tell() without writing them to the stream at all, and then write remaining bytes to the stream.
> 
> I prefer the simpler solution of the former, absent any customer demand for the more complex solution. A user deliberately passing in an un-seekable stream is a rare case and we don't need to write a bunch of hard to test code to support it unless it's obvious people are really doing that.

That makes sense, thanks! The complex solution would be difficult to test. I'll add a clear message if seek() errors out.